### PR TITLE
FIX: Clearing a persistent list does not mark it as changed.

### PR DESCRIPTION
**Pull request by lungj**
The inherited `clear()` method in UserList can modify the `PersistentList` without causing the object to be marked as changed. This patch causes the `clear()` method to mark the object as changed.

### CODE REVIEW

**Comment by jamadden** [comment]
Thanks for the pull request! That's a great find.

It's missing just a few small things. It'll need a test case added to `persistent/tests/test_list.py` (and that `# pragma: no cover` removed, I think), and it'll need a change note added to `CHANGES.rst`. Let me know if I can assist with that.

We may also need a contributor agreement to be signed. I'm not 100% sure what the status of that is, though. @tseaver may be able to help out.

**Comment by lungj** [comment]
Wasn't sure whether to change the tests file to import `PYTHON2` globally in order to only run `test_clear` if not running with Python 2. Made the test run always but only execute the body if not in Python 2.

**Comment by jamadden** [comment]
> Wasn't sure whether to change the tests file to import PYTHON2 globally in order to only run test_clear if not running with Python 2. Made the test run always but only execute the body if not in Python 2.

Great question. I tend to be of the opinion that it's generally best to do specific feature tests rather than rely on preset broad conditions. So I would probably write that test something like this:

```python
def test_clear(self):
    inst = self._makeOne([1])
    if not hasattr(inst, 'clear'):
        raise unittest.SkipTest("PersistentList has no clear() method")
    self.assertFalse(inst._p_changed)
    inst.clear()
    self.assertEqual(inst, [])
    self.assertTrue(inst._p_changed)   
```

That makes it easier to add an implementation of `clear()` in the future on versions where it's missing without accidentally losing test coverage.

Now that I think of it, that probably wouldn't be a bad thing in and of itself (in a following PR):
```python
class PersistentList(UserList):
    __super_clear = (UserList.clear 
        if hasattr(UserList, 'clear') 
        else lambda inst: del inst[:])

    def clear(self):
         self.__super_clear()
        self._p_changed = 1
```

**Comment by lungj** [comment]
The addition of a `clear()` method is actually what I wrote at first :)

File: persistent/list.py [hunk 1]
```diff
@@ -37,14 +46,23 @@ class PersistentList(UserList, persistent.Persistent):
     __super_reverse = UserList.reverse
     __super_sort = UserList.sort
     __super_extend = UserList.extend
+    __super_clear = (
+        UserList.clear
+        if hasattr(UserList, 'clear')
+        else lambda inst: inst.__delitem__(slice(None, None, None))
```
Comment (tseaver) [inline persistent/list.py]: While not use the `_SLICE_ALL` defined above?

File: persistent/list.py [hunk 2]
```diff
@@ -55,8 +73,19 @@ def __setslice__(self, i, j, other):
...
+            # On Python 2, the slice dunders, like ``__delslice__``,
+            # are documented as getting ``(0, sys.maxsize)`` as the
+            # arguments for list[:]. Prior to 2.7.10, it was
+            # incorrectly documented as ``(0, sys.maxint)``, but that
+            # usually worked because ``sys.maxsize`` and
+            # ``sys.maxint`` are usually the same. But on 64-bit
+            # Windown, where ``sizeof(int) == sizeof(long) == 4``,
+            # they're different (``sys.maxsize`` is bigger). See
+            # https://bugs.python.org/issue23645
+            needs_changed = i == 0 and j == sys.maxsize and bool(self)
```
Comment (tseaver) [inline persistent/list.py]: I think this change now masks "normal" list slice deetions (e.g., `del user_list[3:4]`) as being marked changed.

File: persistent/tests/test_list.py [hunk 3]
```diff
@@ -268,13 +268,22 @@ def test_setslice(self):
...
         self.assertTrue(inst._p_changed)
 
-    def test_delslice(self):
+    def test_delslice_nonempty(self):
         inst = self._makeOne([1, 2, 3])
         self.assertFalse(inst._p_changed)
         self.assertEqual(inst, [1, 2, 3])
         del inst[:]
+        self.assertEqual(inst, [])
         self.assertTrue(inst._p_changed)
```
Comment (tseaver) [inline persistent/tests/test_list.py]: We need a test here for a non-empty list with a non-"all" slice.

**Comment by jamadden** [comment]
Thanks for the review. I will investigate tomorrow.

File: persistent/tests/test_list.py (hunk 3 above)
Comment (jamadden) [reply persistent/tests/test_list.py]: Tests and fixes added in https://github.com/zopefoundation/persistent/pull/128
